corehq/apps/users/tests/test_device_rate_limiter.py

```diff
@@ -0,0 +1,62 @@
+from django.test import SimpleTestCase, override_settings
+from freezegun import freeze_time
+
+from corehq.apps.cloudcare.const import DEVICE_ID as CLOUDCARE_DEVICE_ID
+from corehq.apps.users.device_rate_limiter import device_rate_limiter
+from corehq.tests.pytest_plugins.reusedb import clear_redis
+from corehq.util.test_utils import flag_enabled
+
+
+@freeze_time("2024-12-10 12:05:43")
+@override_settings(DEVICE_LIMIT_PER_USER=1)
+class TestDeviceRateLimiter(SimpleTestCase):
+
+    domain = 'device-rate-limit-test'
+
+    def setUp(self):
+        self.addCleanup(clear_redis)
+
+    def test_allowed_if_no_devices_have_been_used_yet(self):
+        self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
+
+    @override_settings(DEVICE_LIMIT_PER_USER=2)
+    def test_allowed_if_device_count_is_under_limit(self):
+        device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
+        self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
+
+    def test_rate_limited_if_device_count_exceeds_limit(self):
+        device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
+        self.assertTrue(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
+
+    def test_allowed_if_device_has_already_been_used(self):
+        device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
+        self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id'))
+
+    def test_allowed_if_different_user(self):
+        device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
+        self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'new-user-id', 'existing-device-id'))
+
+    def test_allowed_after_waiting_one_minute(self):
+        with freeze_time("2024-12-10 12:05:43") as frozen_time:
+            device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'existing-device-id')
+            self.assertTrue(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
+            frozen_time.move_to("2024-12-10 12:06:15")
+            self.assertFalse(device_rate_limiter.rate_limit_device(self.domain, 'user-id', 'new-device-id'))
```

Comment:
I'm not understanding how this works. Does `frozen_time.move_to(...)` affect the redis concept of time? Also, the time window was only moved 32 seconds past 12:05:43, so it seems like it should not have reset yet even if it did affect the redis TTL.

Reply:
Perhaps my comment in https://github.com/dimagi/commcare-hq/pull/35613#discussion_r1924437377 sheds light on this as well, but `_get_redis_key` bakes the timestamp down to the current minute into it, so we don't need redis' concept of time to change, we just need to move to a new fixed window (i.e. the next minute). This is why I chose seemingly random times between these two minutes that were less than a minute apart as well.